File: UAS pemograman berorientasi objek/TransGo/TransGo/oop_03_pewarisan.py
# =============================================================================
# FILE 3 / 4 — PEWARISAN (Inheritance)
# =============================================================================
# Konsep: subclass mewarisi seluruh atribut & method dari parent class
# lewat tanda kurung `class Anak(Induk):`, supaya kode yang sama (id,
# saldo, top_up, dst) tidak perlu ditulis ulang di setiap class
# turunan. Subclass lalu tinggal menambah atribut baru yang spesifik
# untuknya, dan MENIMPA (override) method yang perlu berperilaku beda
# — itulah yang jadi bibit POLIMORFISME di file 4.
#
# File ini WAJIB di-import SETELAH oop_01_abstraksi.py, karena:
#   - Penumpang & Pengemudi butuh class Pengguna sudah ada.
#   - Motor, Mobil, MobilPremium butuh class Kendaraan sudah ada.
# =============================================================================

from datetime import datetime

from oop_01_abstraksi import Pengguna, Kendaraan, AutentikasiError


# -----------------------------------------------------------------------
# Penumpang & Pengemudi — keduanya turunan dari Pengguna, mewarisi
# id, saldo, top_up(), kurangi_saldo(), dst tanpa perlu menulis ulang.
# Yang di-override: login(), get_profil(), role() — inilah
# POLIMORFISME-nya (lihat oop_04_polimorfisme.py).
# -----------------------------------------------------------------------
class Penumpang(Pengguna):
    def __init__(self, nama, email, password):
        super().__init__(nama, email, password)  # panggil constructor Pengguna
        self.riwayat_pesanan = []

    def login(self, password_input):
        if not self.cek_password(password_input):
            raise AutentikasiError("Email atau password penumpang salah.")
        return True

    def get_profil(self):
        return {
            "id": self.id, "nama": self.nama, "email": self.email, "saldo": self._saldo,
            "role": self.role(), "totalPesanan": len(self.riwayat_pesanan),
        }

    def role(self):
        return "penumpang"

    def tambah_riwayat(self, pesanan_id):
        self.riwayat_pesanan.append(pesanan_id)

    def serialize(self):
        data = super().serialize()
        data["riwayatPesanan"] = self.riwayat_pesanan
        return data

    @staticmethod
    def deserialize(data):
        u = Penumpang(data["nama"], data["email"], data["password"])
        u.id = data["id"]
        u._saldo = data["saldo"]
        u.dibuat_pada = datetime.fromisoformat(data["dibuatPada"])
        u.riwayat_pesanan = data.get("riwayatPesanan", [])
        return u


class Pengemudi(Pengguna):
    def __init__(self, nama, email, password, jenis_kendaraan="Motor"):
        super().__init__(nama, email, password)  # panggil constructor Pengguna
        self.jenis_kendaraan = jenis_kendaraan
        self.status_tersedia = True
        self.riwayat_perjalanan = []

    def login(self, password_input):
        if not self.cek_password(password_input):
            raise AutentikasiError("Email atau password pengemudi salah.")
        return True

    def get_profil(self):
        return {
            "id": self.id, "nama": self.nama, "email": self.email, "saldo": self._saldo,
            "role": self.role(), "jenisKendaraan": self.jenis_kendaraan,
            "totalPerjalanan": len(self.riwayat_perjalanan),
        }

    def role(self):
        return "pengemudi"

    def tambah_riwayat(self, pesanan_id):
        self.riwayat_perjalanan.append(pesanan_id)

    def serialize(self):
        data = super().serialize()
        data.update({
            "jenisKendaraan": self.jenis_kendaraan,
            "statusTersedia": self.status_tersedia,
            "riwayatPerjalanan": self.riwayat_perjalanan,
        })
        return data

    @staticmethod
    def deserialize(data):
        u = Pengemudi(data["nama"], data["email"], data["password"], data.get("jenisKendaraan", "Motor"))
        u.id = data["id"]
        u._saldo = data["saldo"]
        u.dibuat_pada = datetime.fromisoformat(data["dibuatPada"])
        u.status_tersedia = data.get("statusTersedia", True)
        u.riwayat_perjalanan = data.get("riwayatPerjalanan", [])
        return u


def deserialize_user(data):
    return Pengemudi.deserialize(data) if data.get("role") == "pengemudi" else Penumpang.deserialize(data)


# -----------------------------------------------------------------------
# Motor, Mobil, MobilPremium — ketiganya turunan dari Kendaraan,
# mewarisi id, plat_nomor, kapasitas, set_tersedia(), dst. Yang
# di-override cuma hitung_tarif() & get_jenis(), masing-masing dengan
# rumus/label berbeda — inilah POLIMORFISME-nya (lihat
# oop_04_polimorfisme.py).
# -----------------------------------------------------------------------
class Motor(Kendaraan):
    def hitung_tarif(self, jarak_km):
        return 2000 + 800 * jarak_km

    def get_jenis(self):
        return "Motor"


class Mobil(Kendaraan):
    def hitung_tarif(self, jarak_km):
        return 5000 + 1500 * jarak_km

    def get_jenis(self):
        return "Mobil"


class MobilPremium(Kendaraan):
    def hitung_tarif(self, jarak_km):
        return (10000 + 3000 * jarak_km) * 1.15

    def get_jenis(self):
        return "Mobil Premium"


def deserialize_kendaraan(data):
    jenis = data["jenis"]
    if jenis == "Motor":
        k = Motor(data["platNomor"], data["namaKendaraan"], data["kapasitas"])
    elif jenis == "Mobil":
        k = Mobil(data["platNomor"], data["namaKendaraan"], data["kapasitas"])
    else:
        k = MobilPremium(data["platNomor"], data["namaKendaraan"], data["kapasitas"])
    k.id = data["id"]
    k.tersedia = data["tersedia"]
    return k

File: UAS pemograman berorientasi objek/TransGo/TransGo/oop-03-pewarisan.js
/* =========================================================================
   FILE 3 / 4 — PEWARISAN (Inheritance)
   =========================================================================
   Konsep: subclass mewarisi seluruh atribut & method dari parent class
   lewat kata kunci `extends`, supaya kode yang sama (id, saldo, topUp,
   dst) tidak perlu ditulis ulang di setiap class turunan. Subclass lalu
   tinggal menambah atribut baru yang spesifik untuknya, dan MENIMPA
   (override) method yang perlu berperilaku beda — itulah yang jadi bibit
   POLIMORFISME di file 4.

   File ini WAJIB dimuat SETELAH oop-01-abstraksi.js, karena:
     - Penumpang & Pengemudi butuh class Pengguna sudah ada.
     - Motor, Mobil, MobilPremium butuh class Kendaraan sudah ada.
   ========================================================================= */

/* ---------------------------------------------------------------
   Penumpang & Pengemudi — keduanya `extends Pengguna`, mewarisi
   id, saldo, topUp(), kurangiSaldo(), dst tanpa perlu menulis
   ulang. Yang di-override: login(), getProfil(), role() — inilah
   POLIMORFISME-nya (lihat oop-04-polimorfisme.js).
   --------------------------------------------------------------- */
class Penumpang extends Pengguna {
  constructor(nama, email, password) {
    super(nama, email, password); // panggil constructor Pengguna
    this.riwayatPesanan = [];
  }
  login(passwordInput) {
    if (!this.cekPassword(passwordInput)) throw new AutentikasiError("Email atau password penumpang salah.");
    return true;
  }
  getProfil() {
    return { id: this.id, nama: this.nama, email: this.email, saldo: this._saldo, role: this.role(), totalPesanan: this.riwayatPesanan.length };
  }
  role() { return "penumpang"; }
  tambahRiwayat(pesananId) { this.riwayatPesanan.push(pesananId); }
  serialize() { return { ...super.serialize(), riwayatPesanan: this.riwayatPesanan }; }
  static deserialize(data) {
    const u = new Penumpang(data.nama, data.email, data.password);
    u.id = data.id;
    u._saldo = data.saldo;
    u.dibuatPada = new Date(data.dibuatPada);
    u.riwayatPesanan = data.riwayatPesanan || [];
    return u;
  }
}

class Pengemudi extends Pengguna {
  constructor(nama, email, password, jenisKendaraan = "Motor") {
    super(nama, email, password); // panggil constructor Pengguna
    this.jenisKendaraan = jenisKendaraan;
    this.statusTersedia = true;
    this.riwayatPerjalanan = [];
  }
  login(passwordInput) {
    if (!this.cekPassword(passwordInput)) throw new AutentikasiError("Email atau password pengemudi salah.");
    return true;
  }
  getProfil() {
    return { id: this.id, nama: this.nama, email: this.email, saldo: this._saldo, role: this.role(), jenisKendaraan: this.jenisKendaraan, totalPerjalanan: this.riwayatPerjalanan.length };
  }
  role() { return "pengemudi"; }
  tambahRiwayat(pesananId) { this.riwayatPerjalanan.push(pesananId); }
  serialize() {
    return { ...super.serialize(), jenisKendaraan: this.jenisKendaraan, statusTersedia: this.statusTersedia, riwayatPerjalanan: this.riwayatPerjalanan };
  }
  static deserialize(data) {
    const u = new Pengemudi(data.nama, data.email, data.password, data.jenisKendaraan);
    u.id = data.id;
    u._saldo = data.saldo;
    u.dibuatPada = new Date(data.dibuatPada);
    u.statusTersedia = data.statusTersedia;
    u.riwayatPerjalanan = data.riwayatPerjalanan || [];
    return u;
  }
}

function deserializeUser(data) {
  return data.role === "pengemudi" ? Pengemudi.deserialize(data) : Penumpang.deserialize(data);
}

/* ---------------------------------------------------------------
   Motor, Mobil, MobilPremium — ketiganya `extends Kendaraan`,
   mewarisi id, platNomor, kapasitas, setTersedia(), dst. Yang
   di-override cuma hitungTarif() & getJenis(), masing-masing
   dengan rumus/label berbeda — inilah POLIMORFISME-nya (lihat
   oop-04-polimorfisme.js).
   --------------------------------------------------------------- */
class Motor extends Kendaraan {
  hitungTarif(jarakKm) { return 2000 + 800 * jarakKm; }
  getJenis() { return "Motor"; }
}
class Mobil extends Kendaraan {
  hitungTarif(jarakKm) { return 5000 + 1500 * jarakKm; }
  getJenis() { return "Mobil"; }
}
class MobilPremium extends Kendaraan {
  hitungTarif(jarakKm) { return (10000 + 3000 * jarakKm) * 1.15; }
  getJenis() { return "Mobil Premium"; }
}

function deserializeKendaraan(data) {
  let k;
  if (data.jenis === "Motor") k = new Motor(data.platNomor, data.namaKendaraan, data.kapasitas);
  else if (data.jenis === "Mobil") k = new Mobil(data.platNomor, data.namaKendaraan, data.kapasitas);
  else k = new MobilPremium(data.platNomor, data.namaKendaraan, data.kapasitas);
  k.id = data.id;
  k.tersedia = data.tersedia;
  return k;
}
